### PR TITLE
perf(python): Add `__slots__` to most Polars classes

### PR DESCRIPTION
**Pull request by Object905**
Closes https://github.com/pola-rs/polars/issues/13198

In this PR are "safe" additions of `__slots__`. Mostly for classes that are not usually subclassed by users -  `pl.Expr`, its subclasses and build in namespaces.
`polars.datatypes.classes` also really good candidates, since they are instantiated now, but not included in this PR (yet?). 

Adding `__slots__` for LazyFrame (and `DataFrame`/`Series`) turned out to be a breaking change because of subclassing.
Using `ldf.__class__ = SubClassedLazyFrame`, like in [test_preservation_of_subclasses](https://github.com/pola-rs/polars/blob/bb0081236ec05af304ec36dcf38673b36fdad706/py-polars/tests/unit/test_lazy.py#L1105) is an error because of different object layouts. And even if subclass were to define `__slots__` - they can only be empty, without additional attributes on instance to keep this assigment valid.
That can't be solved without specific conversion mechanism to subclasses to properly create a new instance (like `pl.LazyFrame.as_subclass(SubClassedLazyFrame, *args, **kwargs)`) where `_ldf` attribute itself would be passed/assigned to new instance of subclass.
Also using `ldf.__class__ = SubClassedLazyFrame` a dirty hack anyway and if polars were to properly support subclassing of frames, I think something similar should be in place - then `__slots__` on them can be added safely.

Deeper description about `__slots__` and benefits\downsides can be found [here](https://stackoverflow.com/questions/472000/usage-of-slots).
Also in issue were not mentioned memory usage benefits aside from slightly faster attribute access.
 ([sqlalchemy](https://docs.sqlalchemy.org/en/20/changelog/migration_10.html#significant-improvements-in-structural-memory-use) claimed 46% less memory usage for it's internal structures). Even though in latest python versions this gap should be smaller, due to more compact dict.

### CODE REVIEW

File: py-polars/polars/api.py
Comment (Object905) [inline py-polars/polars/api.py:46]: I think this were not required even before, but it served as "cache" on instance of expression.
Now it gives error because instance does not have __dict__ and _accessor goes straight to descriptor - and it's read only.

But since expressions are frequently recreated, then this "cache" were only working on chained namespace access like
```pl.col("foo").namespace_foo.func().namespace_foo.other_func()```

**Comment by stinodego** [comment]
Thanks for the PR. Could you provide some data on the performance impact of this change? I think I'm in favor of this change, but it would be nice to have some numbers to support it.

About subclassing: we don't support this. That test is from a while ago and we no longer guarantee preserving the class on calling a method.

**Comment by Object905** [comment]
It's hard to measure exactly, because of variety of possible use cases. But speedup is small, which is anticipated.

In my production use case constructing pl.Expr and pl.LazyFrame with all the glue takes ~20% of request runtime, while collecting them only about 10% because dataframes are pretty small (100-2000 rows). 
And given that this change is sadly close to noise.

Anyway here is some synthetic bench.

upstream
```
In [5]: %timeit -n 5000 pl.when(pl.col("foo") == 1).then("bar").when(pl.col("foo") == 2).then(pl.col("bar") / 2).when(pl.col("bar") == p
   ...: l.col("foo")).then("baz").otherwise(pl.col("bar").list.eval(pl.element().str.starts_with("ni")))
34.4 µs ± 480 ns per loop (mean ± std. dev. of 7 runs, 5,000 loops each)
```
```
In [2]: %timeit -n 10000 -r 100 pl.col("foo").alias("bar")
1.41 µs ± 43.8 ns per loop (mean ± std. dev. of 100 runs, 10,000 loops each)
```
```
In [3]: sys.getsizeof(pl.col.foo)
Out[3]: 48
```

this PR
```
In [7]: %timeit -n 5000 pl.when(pl.col("foo") == 1).then("bar").when(pl.col("foo") == 2).then(pl.col("bar") / 2).when(pl.col("bar") == p
   ...: l.col("foo")).then("baz").otherwise(pl.col("bar").list.eval(pl.element().str.starts_with("ni")))
33.5 µs ± 505 ns per loop (mean ± std. dev. of 7 runs, 5,000 loops each)
```
```
In [3]: %timeit -n 10000 -r 100 pl.col("foo").alias("bar")
1.33 µs ± 41.3 ns per loop (mean ± std. dev. of 100 runs, 10,000 loops each)
```
```
In [3]: sys.getsizeof(pl.col.foo)
Out[3]: 40
```

Also docs are failing to build because of this, because sphinx tries to get `__dict__` on namespaces to check for classmethod/staticmethod.
If it's acceptable, I will add workaround for docs and `__slots__` to other wrappers (LazyFrame/DataFrame/Series) later, otherwise might as well close this PR.

File: py-polars/polars/datatypes/classes.py
Comment (Object905) [inline py-polars/polars/datatypes/classes.py:1]: Turned out that some of parametric tests and py-polars rust code were depending on default values of None in their attributes in Datetime/Duration/List, so I left them as is.

**Comment by Object905** [comment]
Ready to be merged or closed :neutral_face:

**Comment by alexander-beedie** [comment]
> Ready to be merged or closed 😐

If you can rebase against the latest code I'll see about reviewing this 👌

**Comment by Object905** [comment]
@alexander-beedie
Sorry for disappearance. Rebased main.

**Comment by deanm0000** [comment]
Does this prevent monkey patching?

I use `pl.Expr.some_func = lambda x: x #some cool thing` from time to time and it seems `__slots__` would prevent that.

Also, what about this syntax `pl.col.a` instead of `pl.col('a')`?

**Comment by Object905** [comment]
@deanm0000 
`pl.Expr.some_func = ...` modifies class `__dict__`, which is still in place. But this would prevent 
```
instance = pl.col("a")
instance.some_func = ...
```

`pl.col.a` works, because its implemented with __getattr__.

File: py-polars/polars/api.py
Comment (stinodego) [reply py-polars/polars/api.py:46]: I'd appreciate if @alexander-beedie could sign off on this change before this gets merged. I don't really see exactly what is going on here.

File: py-polars/polars/expr/array.py
Comment (stinodego) [inline py-polars/polars/expr/array.py:20]: I'm not a fan of this construction. I would expect the `sphinx_accessor` to take care of this, but it doesn't. We should fix that directly rather than introducing this check everywhere.

File: py-polars/polars/datatypes/classes.py
Comment (stinodego) [inline py-polars/polars/datatypes/classes.py:382]: Nit: remove the blank line here to be consistent with other slots implementations:
```suggestion
    __slots__ = ("precision", "scale")
```

**Review by stinodego — CHANGES_REQUESTED** [review]
Thanks, I left some comments.

I think `Series` and `DataFrame` can also get a `__slots__` attribute. We do not guarantee subclass preservation, so I removed the related tests in another PR.

The way I see it, the biggest blocker for this PR for now is the Sphinx issue, which we should be able to address. Or we could add slots for these namespace classes later.

**Comment by alexander-beedie** [comment]
> @alexander-beedie
> Sorry for disappearance. Rebased main.

Thx; will look at the weekend

File: py-polars/polars/expr/array.py
Comment (Object905) [reply py-polars/polars/expr/array.py:20]: Agree on this. Will try to look into this.
I've tried monkey-patching, but dynamically  adding `__dict__` is way harder when `__slots__` are in place, because it modifies object layout.
Problem seems to be that autosummary makes objects into "modules" and sphinx expects module objects to have dict, so exception thrown from sphinx itself.
Maybe will try to make subclass with `__dict__` dynamically.

Comment (Object905) [reply py-polars/polars/expr/array.py:20]: Created pull request to handle this https://github.com/xarray-contrib/sphinx-autosummary-accessors/pull/124

Comment (stinodego) [reply py-polars/polars/expr/array.py:20]: That's great! They have not released in a year though, so I don't expect this to be picked up quickly.

I'd suggest removing the namespace slots for now - we can add them again when the issue is resolved.

**Review by stinodego — APPROVED** [review]
I removed the namespace slots for now. Let's merge this and see if we get any user feedback on this. If not, that would be a good thing :)

**Comment by lorentzenchr** [comment]
@stinodego May I ask what the decision criteria for inclusion finally was? I'm really just curious because I considered something similar in scikit-learn and measurement showed no performance improvement at all, quite similar to https://github.com/pola-rs/polars/pull/13236#issuecomment-1869714451 did neither.

**Comment by stinodego** [comment]
> @stinodego May I ask what the decision criteria for inclusion finally was? I'm really just curious because I considered something similar in scikit-learn and measurement showed no performance improvement at all, quite similar to [#13236 (comment)](https://github.com/pola-rs/polars/pull/13236#issuecomment-1869714451) did neither.

I like that this clearly documents/enforces what the instance variables are. That is the most important thing for me.

I found out later that this changes the behavior of `getattr` and I'm not sure I'm happy with that. But besides that, I don't see any big downsides so might as well give this a try. There may be some users out there that hit a specific case where the potential performance benefits actually help them.

```python
class A:
    __slots__ = ("x",)

    def __init__(self) -> None:
        self.x = 1

print(getattr(A, "x", None))
# <member 'x' of 'A' objects>       <-- without slots this would print None
```

**Comment by lorentzenchr** [comment]
@stinodego Thanks for the explanation.

File: py-polars/polars/api.py
Comment (stinodego) [reply py-polars/polars/api.py:46]: Turns out this did have user impact:
https://github.com/pola-rs/polars/issues/14851

For this reason, and some related issues popping up, I will be reverting this PR.

**Pull request merged**
